server/athenian/api/align/queries/goals.py

```diff
@@ -145,3 +189,30 @@ def _team_tree_to_team_goal_tree(
         for child in team_tree.children
     ]
     return TeamGoalTree(team_tree, goal_value, children)
+
+
+class _TeamFilterFromTeamGoalRows:
+    def __init__(self, rows: Iterable[Row]):
+        self._team_ids = {row[TeamGoal.team_id.name] for row in rows}
+
+    def __call__(self, team_id: int):
+        return team_id in self._team_ids
+
+
+def _team_tree_prune_empty_branches(
+    team_tree: TeamTree,
+    keep_team_fn: Callable[[int], bool],
+) -> Optional[TeamTree]:
+    """Remove unwanted teams from a TeamTree.
+
+    Empty branches are pruned.
+    """
+    kept_children = [
+        kept
+        for kept in (_team_tree_prune_empty_branches(c, keep_team_fn) for c in team_tree.children)
+        if kept is not None
+    ]
```

Comment:
```suggestion
    kept_children = [
        kept
        for c in team_tree.children
        if (kept := _team_tree_prune_empty_branches(c, keep_team_fn)) is not None
    ]
```
without the second iterator